src/libs/elektra/key.c

```diff
@@ -423,43 +399,37 @@ Key * keyCopy (Key * dest, const Key * source, elektraCopyFlags flags)
 		}
 		else
 		{
-			dest->meta = 0;
+			dest->meta = NULL;
 		}
 	}
 
 	// successful, now do the irreversible stuff: we obviously modified dest
 	set_bit (dest->flags, KEY_FLAG_SYNC);
 
 	// free old resources of destination
-	if (test_bit (flags, KEY_CP_NAME) && !test_bit (orig.flags, KEY_FLAG_MMAP_KEY)) elektraFree (orig.key);
-	if (test_bit (flags, KEY_CP_NAME) && !test_bit (orig.flags, KEY_FLAG_MMAP_KEY)) elektraFree (orig.ukey);
-	if (test_bit (flags, KEY_CP_VALUE) && !test_bit (orig.flags, KEY_FLAG_MMAP_DATA))
-	{
-		elektraFree (orig.data.v);
-	}
-	else if (test_bit (flags, KEY_CP_STRING) && !test_bit (orig.flags, KEY_FLAG_MMAP_DATA))
-	{
-		elektraFree (orig.data.c);
-	}
-
+	keyNameRefDecAndDel (orig.keyName, !test_bit (orig.flags, KEY_FLAG_MMAP_KEY));
+	keyDataRefDecAndDel (orig.keyData, !test_bit (orig.flags, KEY_FLAG_MMAP_DATA));
 	if (test_bit (flags, KEY_CP_META)) ksDel (orig.meta);
 
 	return dest;
 
 memerror:
-	elektraFree (dest->key);
-	elektraFree (dest->data.v);
+	keyNameRefDecAndDel (dest->keyName, !test_bit (dest->flags, KEY_FLAG_MMAP_KEY));
+	keyDataRefDecAndDel (dest->keyData, !test_bit (dest->flags, KEY_FLAG_MMAP_DATA));
 	ksDel (dest->meta);
```

Comment:
We should document (and add to the release notes) that `keyCopy` now only allocates memory, if `KEY_CP_META` (or `KEY_CP_ALL`) is used. A `keyDup` of course still allocates memory (because it calls `keyNew`), but still knowing that `keyCopy` can work without allocations can be useful, e.g. reusing a `Key` for `keyCopy` instead of doing many `keyDup` has now much more effect.

Reply:
Speaking of `keyDup`... currently, it is implemented as a `static inline` function in the header. This is fine, I guess. However, now with COW, we could make it more efficient. 

Currently, it always allocates a name (because of the public API of `keyName` will always allocate a name). With COW we don't need to do that anymore. But that would require converting it to a normal function and not `static inline`.

Reply:
Hm, yeah it could be better. But I'd leave it for now. I believe the unnecessary allocation should be solved via a more generally useful optimization and not something specific to `keyDup`.